modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+    over a collection of elements in a structured manner. It is intended to
+    replace the use of arrays to perform "vector-like" operations. Unlike a
+    stack, the List type also supports inserts or removals into the middle of
+    the List. The List type is close in spirit to its Python counterpart, with
+    fast O(1) random access and append operations.
+
+    The List type is not parallel safe by default. For situations in which
+    such protections are desirable, parallel safety can be enabled by setting
+    `parSafe = true` in any List constructor. A List object constructed from
+    another List object inherits the parallel safety mode of that List by
+    default.
+  */
+  record List {
+
+    /* The type of the elements contained in this List. */
+    type eltType;
+
+    /* If `true`, this List will perform parallel safe operations. */
+    param parSafe;
+
+    pragma "no doc"
+    var _size = 0;
+
+    //
+    // Little piece of wizardry I learned from @ronawho in #13104.
+    //
+    pragma "no doc"
+    var _lock$ = if parSafe then new LockWrapper() else none;
+
+    //
+    // Functionally an array of arrays.
+    //
+    pragma "no doc"
+    var _blocks: unmanaged ListBlock(unmanaged ListBlock(eltType)) = nil;
+
+    //
+    // This is the sum of the capacities of all currently allocated ListBlock.
+    //
+    pragma "no doc"
+    var _totalCapacity = 0;
+
+    /*
+      Initializes an empty List containing elements of the given type.
+
+      :arg eltType: The type of the elements of this List.
+      :arg parSafe: If `true`, this List will use parallel safe operations.
+        Is `true` by default.
+    */
+    proc init(type eltType, param parSafe=false) {
+      this.eltType = eltType;
+      this.parSafe = parSafe;
+    }
+
+    /*
+      Initializes a List containing elements that are copy initialized from
+      the elements in the old List.
+
+      :arg other: The List object to initialize from.
+      :arg parSafe: If `true`, this List will use parallel safe operations.
+        Inherits the value from the other List by default.
+    */
+    proc init=(other: List(?t), param parSafe=other.parSafe) {
+      this.eltType = t;
+      this.parSafe = other.parSafe;
+      this.complete();
+      this.extend(other);
+    }
+
+    pragma "no doc"
+    inline proc deinit() {
+      clear();
+    }
+
+    pragma "no doc"
+    inline proc _destroy(ref item: eltType) {
+      chpl__autoDestroy(item);
+    }
+
+    pragma "no doc"
+    proc _move(ref src: ?t, ref dst: t) {
+      __primitive("=", dst, src);
+    }
+
+    //
+    // The downshift/upshift here (subtract 1 from pos, add 1 to result) is
+    // because this computation assumes zero-based indexing, and Chapel uses
+    // one based indexing.
+    //
+    pragma "no doc"
+    inline proc _getBlockIdx(idx: int): int {
+      const pos = idx + _initialCapacity - 1;
+      const result = log2(pos) - log2(_initialCapacity) + 1;
```

Comment:
Huh, I now see that these integer log2 are in Math.chpl... I missed that. Anyway it's probably worth adding a comment to _logBasePow2Help in Math.chpl that it should really use BitOps.clz (which will allow the removal of a loop). However I don't think that BitOps is currently used by the standard module, so if you changed it now, you'd see some tests fail. I think it's something for another PR. (BTW I also avoided using BitOps in Sort.chpl for similar reasons, but it's something we should work past).

Reply:
Can you hard-code (copy-paste?) the logic used in BitOps? That way we don't need to rely on BitOps but we can still use its implementation. Or at least implement the low-level bitwise operation yourself? Since this is a scheme that obtains its index from the most significant bit, perhaps you can reuse logic from my [BitIndexer](https://github.com/LouisJenkinsCS/Chapel-Atomic-Objects/blob/master/BitIndexer.chpl) that I made a while back?

Reply:
Yes, Math.chpl could include declarations like `  extern proc chpl_bitops_parity_32(x: c_uint) : uint(32);
  extern proc chpl_bitops_parity_64(x: c_ulonglong) : uint(64);` and use those. These have runtime support.